From 69feef5a94731581ea0a8d8f9db8a286b12a2670 Mon Sep 17 00:00:00 2001
From: drisspg <drisspguessous@gmail.com>
Date: Thu, 6 Feb 2025 14:36:08 -0800
Subject: [PATCH] Fix broken meta function for flex-attention backwards
 (#146563)

# Summary

Fixes https://github.com/pytorch/pytorch/issues/146377

So what was the original problem: we were codegening a really weird epilogue:

```Python
        # first compute broadcasted dk of shape [Bq, Hkv, KV_LEN, V_HEAD_DIM]
        # then reduce to dk of shape [Bkv, Hkv, KV_LEN, V_HEAD_DIM]
        xindex = index_k + 64*index_n + 64*off_hkv*ks2 + 128*off_zq*ks2
        tl.store(out_ptr0 + (tl.broadcast_to(index_k + 64*index_n + off_hkv*ks1, dk.shape)), dk, mask)
        x5 = (xindex % ks3)
        tmp2 = tl.load(out_ptr0 + (x5 + ks1*off_hkv), mask, eviction_policy='evict_last')
        tl.store(out_ptr1 + (tl.broadcast_to(xindex, dk.shape)), tmp2, mask)
 ```

 This epilogue was writing and then reading from overlapping regions of memory causing a race condition.

 ### Why were we generating this epilgoue

 During the lowering we created a buffer w/ a different size/stride from the expected return strides. I :think this added an implicit node (for doing the permutation of this wrongly strided output to the the expected one from the meta func. The scheduler for some reason thought it was okay to fuse this into the epilogue, tbh I dont know why.

 This fixes the broken meta func and the original repro. I will add a test but it is hard to pop, better than nothing

Pull Request resolved: https://github.com/pytorch/pytorch/pull/146563
Approved by: https://github.com/Chillee
---
 test/inductor/test_flex_attention.py      | 77 +++++++++++++++++++++++
 torch/_higher_order_ops/flex_attention.py | 56 +++++++++++++----
 torch/_inductor/kernel/flex_attention.py  | 49 ++++++++++-----
 3 files changed, 155 insertions(+), 27 deletions(-)

diff --git a/test/inductor/test_flex_attention.py b/test/inductor/test_flex_attention.py
index 0f197b7b0a600..8b4382061b0d9 100644
--- a/test/inductor/test_flex_attention.py
+++ b/test/inductor/test_flex_attention.py
@@ -2507,6 +2507,28 @@ def mask_mod(b, h, q, kv):
         out = func(query, key, value, block_mask=block_mask)
         out.sum().backward()
 
+    @supported_platform
+    def test_strided_backwards(self):
+        shape = (1, 2, 4096, 64)
+        Q = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
+        K = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
+        V = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
+        func = torch.compile(flex_attention, dynamic=True, fullgraph=True)
+
+        K_sliced = K[:, :, :-128]
+        V_sliced = V[:, :, :-128]
+
+        out_eager = flex_attention(Q, K_sliced, V_sliced)
+        out_compiled = func(Q, K_sliced, V_sliced)
+
+        grad = torch.rand_like(out_eager)
+
+        eager_grads = torch.autograd.grad(out_eager, (Q, K, V), grad)
+        compiled_grads = torch.autograd.grad(out_compiled, (Q, K, V), grad)
+
+        for eager, compiled in zip(eager_grads, compiled_grads):
+            torch.testing.assert_close(eager, compiled, atol=9e-3, rtol=0)
+
     @supported_platform
     @common_utils.parametrize("mode", ["eager", "inductor", "paged_attention"])
     @common_utils.parametrize(
@@ -2560,6 +2582,61 @@ def test_flex_attention_stride_ordering(self, mode, permute_order, shape):
             f"Stride order mismatch: out {out_stride_order}, query {query_stride_order}",
         )
 
+    @supported_platform
+    @common_utils.parametrize("mode", ["eager", "inductor"])
+    @common_utils.parametrize(
+        "permute_order",
+        [
+            (0, 1, 2, 3),
+            (1, 0, 2, 3),
+            (0, 2, 1, 3),
+            (2, 0, 1, 3),
+        ],
+    )
+    @common_utils.parametrize("shape", [(2, 5, 128, 16), (4, 2, 64, 16)])
+    def test_flex_attention_backward_stride_ordering(self, mode, permute_order, shape):
+        if TEST_WITH_ROCM:
+            self.skipTest(
+                "ROCM BUG SEE: https://github.com/pytorch/pytorch/issues/140855"
+            )
+        from torch._inductor.ir import get_stride_order
+
+        dtype = torch.float32
+        make_tensor = functools.partial(
+            torch.randn, shape, device="cuda", dtype=dtype, requires_grad=False
+        )
+
+        query, key, value = make_tensor(), make_tensor(), make_tensor()
+        query = query.permute(permute_order)
+        key = key.permute(permute_order)
+        value = value.permute(permute_order)
+
+        query.requires_grad_()
+        key.requires_grad_()
+        value.requires_grad_()
+
+        func = (
+            torch.compile(flex_attention, backend=mode, fullgraph=True)
+            if mode == "inductor"
+            else flex_attention
+        )
+        out = func(query, key, value)
+        grad_output = torch.randn_like(out)
+        out.backward(grad_output)
+
+        for leaf, grad, name in [
+            (query, query.grad, "query"),
+            (key, key.grad, "key"),
+            (value, value.grad, "value"),
+        ]:
+            input_stride_order = get_stride_order(grad.stride())
+            orig_stride_order = get_stride_order(leaf.stride())
+            self.assertEqual(
+                input_stride_order,
+                orig_stride_order,
+                f"Mode: {mode}, Stride order mismatch for {name}: grad {input_stride_order}, input {orig_stride_order}.",
+            )
+
     @supported_platform
     @common_utils.parametrize("compile", [True, False])
     def test_fully_masked_out_rows_0_check(self, compile: bool):
diff --git a/torch/_higher_order_ops/flex_attention.py b/torch/_higher_order_ops/flex_attention.py
index 9b72ca039704c..ca8ccb435c66d 100644
--- a/torch/_higher_order_ops/flex_attention.py
+++ b/torch/_higher_order_ops/flex_attention.py
@@ -780,16 +780,27 @@ def sdpa_dense_backward(
 ]:
     from torch._dynamo._trace_wrapped_higher_order_op import TransformGetItemToIndex
 
-    # Get outputs before calling repeat interleave
+    Bq, _, _, qk_head_dim = query.shape
+    Bkv, Hkv, seq_len_kv, v_head_dim = value.shape
+
+    # Get outputs before calling repeat interleave and permute to input stride orders
     actual_grad_query = torch.empty_like(query)
-    actual_grad_key = torch.empty_like(key)
-    actual_grad_value = torch.empty_like(value)
+
+    actual_grad_key = key.new_empty((Bq, Hkv, seq_len_kv, qk_head_dim))
+    actual_grad_key = _permute_strides(actual_grad_key, key.stride())
+
+    actual_grad_value = value.new_empty((Bq, Hkv, seq_len_kv, v_head_dim))
+    actual_grad_value = _permute_strides(actual_grad_value, value.stride())
 
     def _maybe_new_buffer(
         buffer: Union[torch.Tensor, torch.SymInt, int]
     ) -> Optional[Union[torch.Tensor, torch.SymInt, int]]:
         if isinstance(buffer, torch.Tensor):
-            return torch.empty_like(buffer) if buffer.requires_grad else None
+            return (
+                torch.empty_like(buffer, memory_format=torch.contiguous_format)
+                if buffer.requires_grad
+                else None
+            )
         return buffer
 
     actual_grad_score_mod_captured = [
@@ -884,18 +895,19 @@ def _maybe_new_buffer(
     grad_key = torch.sum(grad_key, 2, keepdim=False)
     grad_value = torch.sum(grad_value, 2, keepdim=False)
 
+    # Fill to correctly strided outputs
+    actual_grad_query.copy_(grad_query)
+    actual_grad_key.copy_(grad_key)
+    actual_grad_value.copy_(grad_value)
+
     if Bq != Bkv:
         assert (
             Bq > 1 and Bkv == 1
         ), f"Bq and Bkv must broadcast. Got Bq={Bq} and Bkv={Bkv}"
 
-        # Reduce DK, DV along broadcasted batches.
-        grad_key = torch.sum(grad_key, 0, keepdim=True)
-        grad_value = torch.sum(grad_value, 0, keepdim=True)
+        actual_grad_key = torch.sum(actual_grad_key, 0, keepdim=True)
+        actual_grad_value = torch.sum(actual_grad_value, 0, keepdim=True)
 
-    actual_grad_query.copy_(grad_query)
-    actual_grad_key.copy_(grad_key)
-    actual_grad_value.copy_(grad_value)
     score_mod_other_buffer_grads = [
         actual_grad.copy_(grad) if isinstance(actual_grad, torch.Tensor) else None
         for actual_grad, grad in zip(
@@ -1146,17 +1158,35 @@ def flex_attention_backward_fake_tensor_mode(
     torch.Tensor, torch.Tensor, torch.Tensor, tuple[Optional[torch.Tensor], ...]
 ]:
     with mode:
+        Bq, _, _, qk_head_dim = query.shape
+        Bkv, Hkv, seq_len_kv, v_head_dim = value.shape
+
         grad_query = torch.empty_like(query)
-        grad_key = torch.empty_like(key)
-        grad_value = torch.empty_like(value)
+        # zeros_and_scatter creates a contiguous zeros tensor -> contiguous_format
         grad_score_mod_captured = tuple(
             [
-                torch.empty_like(buffer)
+                torch.empty_like(buffer, memory_format=torch.contiguous_format)
                 if isinstance(buffer, torch.Tensor) and buffer.requires_grad
                 else None
                 for buffer in score_mod_other_buffers
             ]
         )
+
+        broadcasted_grad_key = key.new_empty((Bq, Hkv, seq_len_kv, qk_head_dim))
+        broadcasted_grad_key = _permute_strides(broadcasted_grad_key, key.stride())
+
+        broadcasted_grad_value = value.new_empty((Bq, Hkv, seq_len_kv, v_head_dim))
+        broadcasted_grad_value = _permute_strides(
+            broadcasted_grad_value, value.stride()
+        )
+
+        if Bq > 1 and Bkv == 1:
+            grad_key = torch.sum(broadcasted_grad_key, dim=0, keepdim=True)
+            grad_value = torch.sum(broadcasted_grad_value, dim=0, keepdim=True)
+        else:
+            grad_key = broadcasted_grad_key
+            grad_value = broadcasted_grad_value
+
         return grad_query, grad_key, grad_value, grad_score_mod_captured
 
 
diff --git a/torch/_inductor/kernel/flex_attention.py b/torch/_inductor/kernel/flex_attention.py
index ce2d5d793216c..be3dc60531121 100644
--- a/torch/_inductor/kernel/flex_attention.py
+++ b/torch/_inductor/kernel/flex_attention.py
@@ -35,6 +35,7 @@
     _full,
     check_and_broadcast_indices,
     empty,
+    empty_like,
     empty_strided,
     expand,
     index_output_size_and_inner_fn,
@@ -72,6 +73,21 @@ def construct_strides(
     return strides
 
 
+def infer_dense_strides(size: Sequence[int], orig_strides: Sequence[int]):
+    """This is a mirror of the same function in aten/src/ATen/ExpandUtils.cpp
+
+    Args:
+        size: The size of the output tensor
+        orig_strides: The strides of the input tensor
+    Returns:
+        List[int]: Dense non-overlapping strides that preserve the input tensor's layout permutation.
+        The returned strides follow the same stride propagation rules as TensorIterator. This matches
+        The behavior of empty_like()
+    """
+    fill_order = get_fill_order(orig_strides, V.graph.sizevars.shape_env)
+    return construct_strides(size, fill_order)
+
+
 def flex_attention_grid(batch_size, q_heads, num_queries, d_model, meta):
     """How is this kernel parallelized?
     We create a grid of (batch_size * num_heads, ceil_div(n_queries, query_block_size), 1)
@@ -986,8 +1002,7 @@ def lower_cpu(
 
     # Construct output layout with strides matching the query.
     out_size = [B, Hq, seq_len_q, v_head_dim]
-    fill_order = get_fill_order(query.get_stride())
-    out_strides = construct_strides(out_size, fill_order)
+    out_strides = infer_dense_strides(out_size, query.get_stride())
 
     layout = FixedLayout(
         query.get_device(),
@@ -1246,8 +1261,7 @@ def flex_attention(
 
     # Construct output layout with strides matching the query.
     out_size = [B, Hq, seq_len_q, v_head_dim]
-    fill_order = get_fill_order(query.get_stride(), V.graph.sizevars.shape_env)
-    out_strides = construct_strides(out_size, fill_order)
+    out_strides = infer_dense_strides(out_size, q_strides)
 
     layout = FixedLayout(
         query.get_device(),
@@ -2317,11 +2331,15 @@ def flex_attention_backward(*args, **kwargs):
 
     mask_graph_buffer = mask_graph_buffer
 
+    # Construct layout with stride order matching K
+    key_size = [Bq, Hkv, seq_len_kv, qk_head_dim]
+    key_strides = infer_dense_strides(key_size, key.get_stride())
+
     layout_broadcasted_k = FixedLayout(
         key.get_device(),
         key.get_dtype(),
-        [Bq, Hkv, seq_len_kv, qk_head_dim],
-        key.get_stride(),
+        key_size,
+        stride=[sympy.sympify(s) for s in key_strides],
     )
 
     # Create delta which will is needed for the bwd's kernel
@@ -2333,15 +2351,18 @@ def flex_attention_backward(*args, **kwargs):
 
     grad_lse_exp2, delta = maybe_realize([grad_lse_exp2, delta])
 
-    # see NOTE:[TritonTemplates with multiple outputs]
-    grad_query = empty_strided(
-        query.get_size(), query.get_stride(), dtype=dtype, device=device
-    )
+    # # see NOTE:[TritonTemplates with multiple outputs]
+    grad_query = empty_like(query)
+
+    # Construct output layout with stride order matching value
+    value_size = [Bq, Hkv, seq_len_kv, v_head_dim]
+    value_strides = infer_dense_strides(value_size, value.get_stride())
+
     broadcasted_grad_value = empty_strided(
-        (Bq, *value.get_size()[1:]),
-        value.get_stride(),
-        dtype=dtype,
-        device=device,
+        value_size,
+        stride=[sympy.sympify(s) for s in value_strides],
+        dtype=value.get_dtype(),
+        device=value.get_device(),
     )
 
     kernel_options.setdefault("SM_SCALE", scale)
